### PR TITLE
Issue #4 - Clean up error handling, make sure disconnect is called co…

### PR DESCRIPTION
**Pull request by nemccarthy**
…rrectly after write. Improve retry logic to also include retrying if the connection attempt fails. Finally logging has been improved.

### CODE REVIEW

File: switchbot/__init__.py [hunk 1]
```diff
@@ -18,53 +22,72 @@
...
+        if not send_success:
+            if retry < 1:
+                _LOGGER.error("Switchbot communication failed. Stopping trying.", exc_info=True)
+            else:
+                _LOGGER.warning("Cannot connect to Switchbot. Retrying (remaining: %d)...", retry)
+                time.sleep(DEFAULT_RETRY_TIMEOUT)
+                return self._sendcommand(key, retry - 1)
+        return send_success
 
     def turn_on(self) -> None:
```
Comment (Danielhiversen) [inline switchbot/__init__.py]: -> bool

File: switchbot/__init__.py [hunk 2]
```diff
@@ -18,53 +22,72 @@
...
         """Turn device on."""
-        return self._sendpacket(ON_KEY)
+        return self._sendcommand(ON_KEY, self._retry_count)
 
     def turn_off(self) -> None:
```
Comment (Danielhiversen) [inline switchbot/__init__.py]: -> bool

File: switchbot/__init__.py [hunk 3]
```diff
@@ -18,53 +22,72 @@
...
         """Turn device off."""
-        return self._sendpacket(OFF_KEY)
+        return self._sendcommand(OFF_KEY, self._retry_count)
 
     def press(self) -> None:
```
Comment (Danielhiversen) [inline switchbot/__init__.py]: -> bool

File: switchbot/__init__.py [hunk 4]
```diff
@@ -18,53 +22,72 @@
...
+                self._device = bluepy.btle.Peripheral(self._mac,
+                                                      bluepy.btle.ADDR_TYPE_RANDOM)
+                _LOGGER.debug("Connected to Switchbot.")
+            except bluepy.btle.BTLEException:
+                _LOGGER.debug("Failed connecting to Switchbot.", exc_info=True)
+                self._device = None
+                raise
 
-    def _connect(self) -> bool:
+    def _disconnect(self):
```
Comment (Danielhiversen) [inline switchbot/__init__.py]: -> None

File: switchbot/__init__.py [hunk 5]
```diff
@@ -18,53 +22,72 @@
 class Switchbot:
     """Representation of a Switchbot."""
 
-    def __init__(self, mac) -> None:
+    def __init__(self, mac, retry_count=DEFAULT_RETRY_COUNT) -> None:
         self._mac = mac
         self._device = None
+        self._retry_count = retry_count
+
+    def _connect(self):
```
Comment (Danielhiversen) [inline switchbot/__init__.py]: -> None

File: switchbot/__init__.py [hunk 6]
```diff
@@ -18,53 +22,72 @@
...
+        if self._device is None:
```
Comment (Danielhiversen) [inline switchbot/__init__.py]:
```

        if self._device is not None:
            return

```

File: switchbot/__init__.py [hunk 7]
```diff
@@ -18,53 +22,72 @@
...
+                                                      bluepy.btle.ADDR_TYPE_RANDOM)
+                _LOGGER.debug("Connected to Switchbot.")
+            except bluepy.btle.BTLEException:
+                _LOGGER.debug("Failed connecting to Switchbot.", exc_info=True)
+                self._device = None
+                raise
 
-    def _connect(self) -> bool:
+    def _disconnect(self):
         if self._device is not None:
```
Comment (Danielhiversen) [inline switchbot/__init__.py]: if self._device is None:
            return

File: switchbot/__init__.py [hunk 8]
```diff
@@ -18,53 +22,72 @@
...
-            if retry < 1 or not self._connect():
-                _LOGGER.error("Can not connect to switchbot.", exc_info=True)
-                return False
-            _LOGGER.warning("Can not connect to switchbot. Retrying")
-            return self._sendpacket(key, retry-1)
-        return True
+            _LOGGER.warning("Error talking to Switchbot.", exc_info=True)
+        finally:
+            self._disconnect()
+        if not send_success:
```
Comment (Danielhiversen) [inline switchbot/__init__.py]:
```
         if retry < 1:
              _LOGGER.error("Switchbot communication failed. Stopping trying.", exc_info=True)
        if send_success:
            return send_success
        _LOGGER.warning("Cannot connect to Switchbot. Retrying (remaining: %d)...", retry)
        time.sleep(DEFAULT_RETRY_TIMEOUT)
        return self._sendcommand(key, retry - 1)
```

**Comment by nemccarthy** [comment]
Thanks for the feedback @Danielhiversen - changes have been made

**Comment by nemccarthy** [comment]
Could we get this released as either 0.6.3 or 0.7.0 and I'll do a PR into HA to have the config option RETRY_COUNT added so users can optionally set this depending on their environment

File: switchbot/__init__.py (hunk 6 above)
Comment (Danielhiversen) [reply switchbot/__init__.py]: Please fix this

**Comment by Danielhiversen** [comment]
Please revert the changes of README.md

**Comment by Danielhiversen** [comment]
Made the fixes here: https://github.com/Danielhiversen/pySwitchbot/pull/7
Is that ok for you?

**Comment by nemccarthy** [comment]
Thanks Daniel. Can we get these merged so we can release a new version and I can commit my homeassistant changes

**Pull request closed without merging**